src/app/containers/RedeemContainer.js

```diff
@@ -1,7 +1,179 @@
-import React from 'react';
+import React, { useState, useCallback } from 'react';
+import PropTypes from 'prop-types';
+import { c } from 'ttag';
+import { useApi, useConfig, Title, Href, VpnLogo, Icon, Button, Input, ErrorZone } from 'react-components';
+import { queryCheckVerificationCode } from 'proton-shared/lib/api/user';
+import { checkSubscription } from 'proton-shared/lib/api/payments';
+import { TOKEN_TYPES } from 'proton-shared/lib/constants';
 
-const RedeemContainer = () => {
-    return <>RedeemContainer</>;
+import GadgetImage from 'design-system/assets/img/pv-images/redeem/gadgets.png';
+import SwissBasedIcon from 'design-system/assets/img/pv-images/redeem/swiss_based.svg';
+import SecureCoreIcon from 'design-system/assets/img/pv-images/redeem/secure_core.svg';
+import NoLogsIcon from 'design-system/assets/img/pv-images/redeem/no_logs.svg';
+import KillSwitchIcon from 'design-system/assets/img/pv-images/redeem/kill_switch.svg';
+import DNSIcon from 'design-system/assets/img/pv-images/redeem/dns.svg';
+import EncryptionIcon from 'design-system/assets/img/pv-images/redeem/encryption.svg';
+
+const RedeemContainer = ({ history }) => {
+    const api = useApi();
+    const { CLIENT_TYPE } = useConfig();
+    const [couponCode, setCouponCode] = useState();
+    const [validationError, setValidationError] = useState();
+
+    const handleCouponCodeChange = useCallback(({ target: { value } }) => {
+        setCouponCode(value);
+    }, []);
+
+    const validateCoupon = useCallback(
+        async (e) => {
+            e.preventDefault();
+
+            try {
+                const {
+                    Plans: [{ ID, Currency, Cycle, Quantity }]
+                } = await api(queryCheckVerificationCode(couponCode, TOKEN_TYPES.COUPON, CLIENT_TYPE));
```

Comment:
You should call the API only when the user submit the code. Not each time the user is typing.

Reply:
It does call the API only when the user submits the code.